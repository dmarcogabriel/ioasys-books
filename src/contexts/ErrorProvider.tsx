import React, {createContext} from 'react';
import {UNAUTHORIZED_STATUS} from '../constants';
import Toast from 'react-native-toast-message';

type ErrorData = {
  message: string;
};

export interface ApiError {
  errors: ErrorData;
}

interface ErrorContextValue {
  handleError: (
    error: ApiError,
    status: number,
    logoutCallback: () => Promise<void>,
  ) => Promise<void>;
}

export const ErrorContext = createContext<ErrorContextValue>({
  handleError: async () => {},
});

interface ErrorProviderProps {
  children: React.ReactNode;
}

export const ErrorProvider = ({children}: ErrorProviderProps): JSX.Element => {
  const handleError = async (
    error: ApiError,
    status: number,
    logoutCallback: () => Promise<void>,
  ): Promise<void> => {
    let errorMessage: string = '';
    if (status === UNAUTHORIZED_STATUS) {
      await logoutCallback();
      errorMessage = 'Sess√£o expirada';
    } else {
      Toast.show({
        position: 'bottom',
        text1: errorMessage || error.errors.message,
        type: 'error',
      });
    }
  };

  return (
    <ErrorContext.Provider value={{handleError}}>
      {children}
    </ErrorContext.Provider>
  );
};
