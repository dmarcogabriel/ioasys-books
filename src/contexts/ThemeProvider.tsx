import React from 'react';
import {ThemeProvider as SCThemeProvider} from 'styled-components/native';

interface ThemeProviderProps {
  children: React.ReactNode;
}

export const ThemeProvider = ({children}: ThemeProviderProps): JSX.Element => {
  const theme = {
    colors: {
      primary: '#B22E6F',
      dark: '#333333',
      light: '#ffffff',
      grey: '#999999',
    },
  };

  return <SCThemeProvider theme={theme}>{children}</SCThemeProvider>;
};
