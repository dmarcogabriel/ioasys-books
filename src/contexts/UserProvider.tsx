import React, {createContext, useState, useEffect, useCallback} from 'react';
import {User} from '../@types/user.interface';
import {storeData, removeData, getData} from '../utils/storage';
import {AUTH_TOKEN_STORAGE_KEY} from '../constants';

interface UserContextValue {
  user: User | null;
  login: (user: User, token: string) => Promise<void>;
  logout: () => Promise<void>;
  token: string | null;
}

export const UserContext = createContext<UserContextValue>({
  user: null,
  token: null,
  login: async () => {},
  logout: async () => {},
});

interface UserProviderProps {
  children: React.ReactNode;
}

export const UserProvider = ({children}: UserProviderProps): JSX.Element => {
  const [user, setUser] = useState<User | null>(null);
  const [token, setToken] = useState<string | null>(null);
  const [isLoading, setIsLoading] = useState(true);

  const login = async (data: User, authToken: string): Promise<void> => {
    setUser(data);
    setToken(authToken);
    await storeData(AUTH_TOKEN_STORAGE_KEY, authToken);
  };

  const logout = async () => {
    await removeData(AUTH_TOKEN_STORAGE_KEY);
    setUser(null);
    setToken(null);
  };

  const loadToken = useCallback(async () => {
    setIsLoading(true);
    try {
      const authToken = await getData<string>(AUTH_TOKEN_STORAGE_KEY);
      if (authToken) {
        setToken(authToken);
      }
    } catch (error) {
      setIsLoading(false);
    }
    setIsLoading(false);
  }, []);

  useEffect(() => {
    loadToken();
  }, [loadToken]);

  return (
    <UserContext.Provider value={{user, token, login, logout}}>
      {!isLoading && children}
    </UserContext.Provider>
  );
};
