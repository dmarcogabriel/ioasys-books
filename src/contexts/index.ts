export * from './UserProvider';
export * from './ThemeProvider';
export * from './ErrorProvider';
