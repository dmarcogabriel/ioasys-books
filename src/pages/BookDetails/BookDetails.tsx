import React, {useCallback, useEffect, useState} from 'react';
import {useRoute, RouteProp} from '@react-navigation/native';
import {
  BookDetailsContainer,
  ContentSection,
  ContentSectionTitle,
  ContentSectionReview,
  BookDetailsImage,
  BookDetailsTitle,
  BookDetailsAuthors,
} from './styles';
import {Book} from '../../types/book.interface';
import {BookInfo} from './components';
import {StatusBar} from 'react-native';
import api from '../../services/api';
import {RootStacksParamsList} from '../../types/rootStacksParamsList.interface';
import {useUser, useErrorHandler} from '../../hooks';

export const BookDetails = () => {
  const {params} = useRoute<RouteProp<RootStacksParamsList, 'BookDetails'>>();
  const [book, setBook] = useState<Book | null>(null);
  const [isLoading, setIsLoading] = useState(true);
  const {logout} = useUser();
  const {handleError} = useErrorHandler();

  const loadBookDetails = useCallback(async () => {
    try {
      setIsLoading(true);
      const {data} = await api.get<Book>(`books/${params.id}`);
      setBook(data);
      setIsLoading(false);
    } catch (err: any) {
      if (err.response && err.response.status) {
        handleError(err.response.data, err.response.status, logout);
      } else {
        setIsLoading(false);
      }
    }
  }, [params.id, logout, handleError]);

  useEffect(() => {
    loadBookDetails();
  }, [loadBookDetails]);

  return (
    <BookDetailsContainer>
      <StatusBar backgroundColor="#ffffff" />
      {!isLoading && book && (
        <>
          <BookDetailsImage
            testID="bookDetailsImage"
            source={{uri: book.imageUrl}}
            resizeMode="cover"
          />
          <BookDetailsTitle testID="bookDetailsTitle">
            {book.title}
          </BookDetailsTitle>
          <BookDetailsAuthors testID="bookDetailsAuthors">
            {book.authors.join(', ')}
          </BookDetailsAuthors>
          <ContentSection>
            <ContentSectionTitle>INFORMAÇÕES</ContentSectionTitle>
            <BookInfo name="Páginas" value={`${book.pageCount} Páginas`} />
            <BookInfo name="Editora" value={book.publisher} />
            <BookInfo name="Publicação" value={book.published} />
            <BookInfo name="Idioma" value={book.language} />
            <BookInfo name="Título Original" value={book.title} />
            <BookInfo name="ISBN-10" value={book.isbn10} />
            <BookInfo name="ISBN-13" value={book.isbn13} />
            <BookInfo name="Categoria" value={book.category} />
          </ContentSection>
          <ContentSection marginBottom="42px">
            <ContentSectionTitle>RESENHA DA EDITORA</ContentSectionTitle>
            <ContentSectionReview testID="bookDetailsReview">
              {book.description}
            </ContentSectionReview>
          </ContentSection>
        </>
      )}
    </BookDetailsContainer>
  );
};
