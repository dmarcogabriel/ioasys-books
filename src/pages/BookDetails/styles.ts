import styled from 'styled-components/native';
import {ThemeProps} from '../../@types/theme.interface';
import {Text} from 'react-native-elements';

export const BookDetailsContainer = styled.ScrollView`
  background: ${({theme}: ThemeProps) => theme.colors.light};
  flex: 1;
  padding: 0 40px;
`;

export const BookDetailsImage = styled.Image`
  height: 351px;
  width: 240px;
  align-self: center;
  margin: 12px 0;
`;

interface ContentSectionProps {
  marginBottom?: string;
}

export const ContentSection = styled.View`
  margin-top: 8px;
  margin-bottom: ${({marginBottom = '8px'}: ContentSectionProps) =>
    marginBottom};
`;

export const ContentSectionTitle = styled(Text)`
  color: ${({theme}: ThemeProps) => theme.colors.dark};
  font-size: 12px;
  font-weight: bold;
  margin-bottom: 16px;
  line-height: 20px;
`;

export const ContentSectionReview = styled(Text)`
  color: ${({theme}: ThemeProps) => theme.colors.grey};
  font-size: 12px;
  line-height: 20px;
`;

export const BookDetailsTitle = styled(Text)`
  color: ${({theme}: ThemeProps) => theme.colors.dark};
  font-size: 28px;
  font-weight: bold;
  margin-top: 12px;
`;

export const BookDetailsAuthors = styled(Text)`
  color: ${({theme}: ThemeProps) => theme.colors.primary};
  font-size: 12px;
  margin-bottom: 24px;
`;
