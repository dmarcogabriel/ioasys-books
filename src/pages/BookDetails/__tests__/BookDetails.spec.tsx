import React from 'react';
import {render} from '@testing-library/react-native';
import {BookDetails} from '../BookDetails';
import {ThemeProvider} from '../../../contexts';
// import {Book} from '../../../@types/book.interface';
import renderer from 'react-test-renderer';

jest.mock('@react-navigation/native', () => ({
  useRoute: () => ({
    params: {id: 'test'},
  }),
}));

const ComponentWrapper = () => (
  <ThemeProvider>
    <BookDetails />
  </ThemeProvider>
);
// const mockBook: Book = {
//   title: 'Test Book',
//   authors: ['Author 1', 'Author 2'],
//   pageCount: 123,
//   publisher: 'Tester',
//   published: 1992,
//   imageUrl: 'test',
//   id: 'test',
//   description: 'Testing',
//   category: 'Test',
//   isbn10: 'test-10',
//   isbn13: 'test-13',
//   language: 'test',
// };

describe('pages/BookDetails', () => {
  it('should render correctly', () => {
    const {getByTestId} = render(<ComponentWrapper />);

    const bookDetailsImage = getByTestId('bookDetailsImage');
    const bookDetailsTitle = getByTestId('bookDetailsTitle');
    const bookDetailsAuthors = getByTestId('bookDetailsAuthors');
    const bookDetailsReview = getByTestId('bookDetailsReview');

    expect(bookDetailsImage).toHaveProp('source', {
      uri: 'https://files-books.ioasys.com.br/Book-0.jpg',
    });
    expect(bookDetailsTitle.children[0]).toBe('A Culpa é das Estrelas');
    expect(bookDetailsAuthors.children[0]).toBe('Jonh Green');
    expect(bookDetailsReview.children[0]).toBe(
      'Hazel foi diagnosticada com câncer aos treze anos e agora, aos dezesseis, sobrevive graças a uma droga revolucionária que detém a metástase em seus pulmões. Ela sabe que sua doença é terminal e passa os dias vendo tevê e lendo Uma aflição imperial, livro cujo autor deixou muitas perguntas sem resposta. ',
    );
  });

  it('should match snapshot', () => {
    expect(renderer.create(<ComponentWrapper />).toJSON()).toMatchSnapshot();
  });
});
