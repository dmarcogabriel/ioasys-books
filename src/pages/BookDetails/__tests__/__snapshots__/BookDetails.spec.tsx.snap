// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages/BookDetails should match snapshot 1`] = `
<RCTScrollView
  style={
    Array [
      Object {
        "backgroundColor": "#ffffff",
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
        "paddingBottom": 0,
        "paddingLeft": 40,
        "paddingRight": 40,
        "paddingTop": 0,
      },
    ]
  }
>
  <View />
</RCTScrollView>
`;
